subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/DefaultFileLockContentionHandler.java

```diff
@@ -23,20 +23,28 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.net.DatagramPacket;
+import java.util.ArrayList;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 import java.util.concurrent.locks.Lock;
 import java.util.concurrent.locks.ReentrantLock;
 
 public class DefaultFileLockContentionHandler implements FileLockContentionHandler, Stoppable {
     private static final Logger LOGGER = LoggerFactory.getLogger(DefaultFileLockContentionHandler.class);
+    private static final int PING_DELAY = 1000;
     private final Lock lock = new ReentrantLock();
     private final Map<Long, Runnable> contendedActions = new HashMap<Long, Runnable>();
+    private final List<Long> contendedActionsRunning = new ArrayList<Long>();
```

Comment:
I put the contended actions state into one structure.

I left `unlocksRequestedFrom` and `unlocksConfirmedFrom` as simple maps from `lockId` to `port` since they only record the incoming messages as they are which are interpreted later in `maybePingOwner`. It's also due to the low level protocol. The packets we send over the socket only contain the lockId and no further context. But I do not want to touch this protocol if not absolutely necessary.